### PR TITLE
Jmap shared upload collection

### PR DESCRIPTION
**Pull request by ksmurchison**
Tests added in https://github.com/cyrusimap/cassandane/pull/112

### CODE REVIEW

File: imap/http_jmap.c [hunk 1]
```diff
@@ -625,6 +625,36 @@ static int jmap_download(struct transaction_t *txn)
     return res;
 }
 
+static int sharedrights_cb(const mbentry_t *mbentry, void *vrock)
```
Comment (rsto) [inline imap/http_jmap.c:628]: Nit: Rename this to has_shared_rw_rights_cb ? This function really just that boolean question, and it's less confusing to than having the same named function to different things in two files.

File: imap/jmap_api.c [hunk 2]
```diff
@@ -3093,6 +3195,14 @@ HIDDEN int jmap_set_sharewith(struct mailbox *mbox,
         }
     }
 
+    if (!r) {
+        /* Set proper access rights on JMAP upload folder */
+        /* XXX This is currently done by brute force.
```
Comment (rsto) [inline imap/jmap_api.c:3200]: This comment might be better placed at the start of set_upload_rights. Without knowing the internals of that function, it is not clear where the "brute force" is coming from here.

**Review by rsto — CHANGES_REQUESTED** [review]
The logic looks sane to me. Found two nits that I think would make the code clearer to read.

**Comment by elliefm** [comment]
Does this need to be backported to 3.4?  The Cassandane test now checks for 3.4 or greater, but 3.4 doesn't currently have this code, so the test runs and fails.  Not sure whether the test is correct and the changes need to be backported, or if they don't and the test's version check is wrong.

It's not clear to me what the big picture of this change is or how it fits into other JMAP stuff that's happened on master (and not 3.4)

**Comment by ksmurchison** [comment]
This should probably be backported if we support sharing of JMAP resources in 3.4, which I assume we do.

**Comment by elliefm** [comment]
The calendar_set_sharewith test previously worked before the specifics for this PR were added to it, so I'm concluding from that that it does support sharing, and therefore this needs to be backported.  Cool, thanks for the clarification, I'll get it done. :)

**Comment by elliefm** [comment]
This is now on 3.4

**Pull request merged**